csharp_dotnetcore/17.Proactive-Messages/Startup.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.BotFramework;
+using Microsoft.Bot.Builder.Integration;
+using Microsoft.Bot.Builder.Integration.AspNet.Core;
+using Microsoft.Bot.Builder.TraceExtensions;
+using Microsoft.Bot.Configuration;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.BotBuilderSamples
+{
+    public class Startup
+    {
+        public Startup(IHostingEnvironment env)
+        {
+            var builder = new ConfigurationBuilder()
+                .SetBasePath(env.ContentRootPath)
+                .AddEnvironmentVariables();
+
+            Configuration = builder.Build();
+        }
+
+        public IConfiguration Configuration { get; }
+
+        public void ConfigureServices(IServiceCollection services)
+        {
+            // Register the proactive bot.
+            services.AddBot<ProactiveBot>(options =>
+            {
+                options.CredentialProvider = new ConfigurationCredentialProvider(Configuration);
+
+                // Set up error handling. (Trace output goes to the Emulator log, but not to the user.)
+                options.OnTurnError = async (context, exception) =>
+                {
+                    await context.TraceActivityAsync("Proactive bot exception", exception);
+                    await context.SendActivityAsync("Sorry, it looks like something went wrong!");
+                };
+
+                // Set up state management middleware.
+                IStorage dataStore = new MemoryStorage();
+                var state = new JobState(dataStore);
+                options.Middleware.Add(state);
```

Comment:
By convention, "JobState" should be "JobStateMiddleware". 
... although I'm not yet clear on why we need a custom middleware component in the Proactive Messaging sample.

Reply:
Job state is not tied to either a user or a conversation. So, I didn't want to use either ConversationState or UserState to manage the job log. (Plus, I guess this is no longer necessarily middleware.)

Reply:
Yep, putting in the State list during startup.